lib/tom_queue/worker_supervisor.rb

```diff
@@ -0,0 +1,81 @@
+module TomQueue
+  class WorkerSupervisor
+
+    attr_accessor :processes, :before_fork, :after_fork
+
+    def initialize
+      @processes = {}
+      @before_fork = -> { }
+      @after_fork = -> { }
+      # TODO setup_signal_handling
+    end
+
+    def supervise(as:, count: 1, &block)
+      count.times do |i|
+        processes["#{as}#{i + 1}"] = block
+      end
+    end
+
+    def setup_signal_handling
+      ["SIGINT","SIGTERM"].each do |signal|
+        Signal.trap(signal) do
+          raise SignalException, "SIGTERM"
+        end
+      end
+    end
+
+    def run
+      process_ids = {}
+      before_fork.call
+      loop do
+        processes_to_start = (processes.keys - process_ids.values)
+
+        processes_to_start.each do |name|
+          $stderr.puts "Starting '#{name}'..."
+          process_id = fork_process(name)
+          process_ids[process_id] = name
+        end
+
+        stopped_process_id, status = wait_for_child_process_to_exit
+
+        if stopped_process_name = process_ids.delete(stopped_process_id)
+          $stderr.puts "#{stopped_process_name} terminated (#{stopped_process_id}) with status #{status}"
+        end
+
+        break if stop_loop?
+        throttle_loop
+      end
+    end
+
+    private
+
+    def fork_process(name)
+      fork do
+        execute_child_process(processes[name])
+      end
+    end
+
+    def execute_child_process(process)
+      after_fork.call
+      process.call
+      stop_child_when_process_finishes
+    end
```

Comment:
The execute_child_process and fork methods are separate just because it made stubbing the #fork method a bit easier